VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,656 @@
+single constraint or a list of constraints separated in turn by an comma "," as
+in ``1.2.3|>=2.0.0,<5.0.0``.
+
+Multiple ``<version-constraint>`` combined with a comma means that **all** these
+constraints must be satisfied for a version to be resolved as contained in this
+range.
+
+Each simple version constraint has this syntax::
+
+    <comparator><version>
+
+The ``<comparator>`` is one of these comparison operators:
+
+- "=": Version equality comparator. It is the default and implied if not
+  present and means that a version must be equal to the provided version.
+  For example: "=1.2.3". It must be omitted in the canonical representation.
+  Equality is based on the equality of two lower-cased and normalized version
+  strings and is typically not versioning scheme-specific, though some
+  scheme such as pypi PEP440 may apply some version string normalization
+  before testing for equality.
+
+- "!=": Version exclusion or inequality comparator. This means a version must
+  not be equal to the provided version and this version must be excluded from
+  the range. For example: "!=1.2.3" means that version "1.2.3" is excluded.
+
+- "<", "<=": Less than or less-or-equal version comparators points to all
+  versions less than or equal to the provided version. For example "<=1.2.3"
+  means less than or equal to "1.2.3". 
+
+- ">", ">=": Greater than or greater-or-equal version comparators points to
+  all versions greater than or equal to the provided version. For example
+  ">=1.2.3" means greater than or equal to "1.2.3".
+
+- The way two version strings are compared using these comparators is defined
+  by the ``<versioning-scheme>``.
+
+- The structure and meaning of a version string such as "1.2.3" is defined by
+  the ``<versioning-scheme>``. For instance, ``semver`` defines three
+  dot-separated segments name major, minor and patch.
+
+- The special star "*" ``<version-constraint>`` matches any version. This star
+  constraint must be used **alone** in a version range, exclusive of any other
+  constraint. For example "vers:deb/\*" resolves to any version of a Debian
+  package.
+
+- The way each of these comparators work when doing a version comparison is
+  specific to a versioning scheme.
+
+
+Examples
+~~~~~~~~~
+
+TODO.
+
+
+Normalized or canonical representation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- A version range specifier contains only printable ASCII letters, digits and
+  punctuation.
+
+- Spaces are not significant and are removed in the canonical form. For example
+  "!=1.2.3" and " ! = 1.2. 3" are equivalent. And so are "1.2.3 & < = 2.0.0" and
+  "1.2.3&<=2.0.0"
+
+- A version range specifier is case-insensitive and lowercase in canonical form.
+
+- The ordering of multiple ``<version-constraint>`` in a range specifier is not
+  significant. The canonical ordering is by sorting these by lexicographical
+  order applied with this two steps approach:
```

Comment:
This sorting ends up being a bit weird, as this sorts first based on the comparator strings and second on the version. Sorting by version is possibly opening a can of worm as  this would be scheme specific.